File: clients/js/packages/chromadb-core/src/embeddings/RunPodEmbeddingFunction.ts
import {
  IEmbeddingFunction,
  EmbeddingFunctionSpace,
} from "./IEmbeddingFunction";
import { validateConfigSchema } from "../schemas/schemaUtils";

declare const process: any;

interface RunPodAPI {
  createEmbedding: (params: {
    model: string;
    input: string[];
    timeout?: number;
  }) => Promise<number[][]>;
}

class RunPodAPIImpl implements RunPodAPI {
  private readonly apiKey: string;
  private readonly endpointId: string;
  private runpod: any;
  private endpoint: any;

  constructor(apiKey: string, endpointId: string, runpodSdk: any) {
    this.apiKey = apiKey;
    this.endpointId = endpointId;

    // Initialize the RunPod SDK
    this.runpod = runpodSdk(this.apiKey);
    this.endpoint = this.runpod.endpoint(this.endpointId);
  }

  public async createEmbedding(params: {
    model: string;
    input: string[];
    timeout?: number;
  }): Promise<number[][]> {
    // Process all documents in parallel for better performance
    const embeddings = await Promise.all(
      params.input.map((document) =>
        this.createSingleEmbedding(document, params.model, params.timeout)
      )
    );

    return embeddings;
  }

  private async createSingleEmbedding(
    document: string,
    model: string,
    timeout?: number
  ): Promise<number[]> {
    try {
      // Prepare the input payload for RunPod
      const inputPayload = {
        input: {
          model: model,
          input: document,
        },
      };

      // Start the async request
      const runRequest = await this.endpoint.run(inputPayload);
      const jobId = runRequest.id;

      // Check initial status
      let status = runRequest.status;

      if (
        status === "FAILED" ||
        status === "CANCELLED" ||
        status === "TIMED_OUT"
      ) {
        throw new Error(
          `RunPod endpoint failed with status '${status}': ${JSON.stringify(
            runRequest,
          )}`,
        );
      }

      let finalStatus;
      if (status === "COMPLETED") {
        finalStatus = await this.endpoint.status(jobId);
      } else {
        const timeoutMs = (timeout || 300) * 1000;
        const startTime = Date.now();
        const pollInterval = 1000;

        while (Date.now() - startTime < timeoutMs) {
          finalStatus = await this.endpoint.status(jobId);

          if (finalStatus.status === "COMPLETED") {
            break;
          } else if (
            finalStatus.status === "FAILED" ||
            finalStatus.status === "CANCELLED" ||
            finalStatus.status === "TIMED_OUT"
          ) {
            throw new Error(
              `RunPod endpoint failed with status '${
                finalStatus.status
              }': ${JSON.stringify(finalStatus)}`,
            );
          }

          await new Promise((resolve) => setTimeout(resolve, pollInterval));
        }

        if (finalStatus?.status !== "COMPLETED") {
          throw new Error(
            `Request timed out after ${
              timeout || 300
            } seconds. Last status: ${finalStatus?.status || "UNKNOWN"}`,
          );
        }
      }

      // Extract embedding from the completed status
      const output = finalStatus?.output;
      if (output && "data" in output) {
        const dataList = output["data"];
        if (dataList.length > 0 && "embedding" in dataList[0]) {
          return dataList[0]["embedding"];
        } else {
          throw new Error(
            `No embedding found in response data: ${JSON.stringify(dataList)}`,
          );
        }
      } else {
        throw new Error(
          `Unexpected output format. Expected 'output.data[0].embedding', got: ${JSON.stringify(
            output,
          )}`,
        );
      }
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : String(error);
      throw new Error(
        `Failed to generate embedding for document: ${errorMessage}`,
      );
    }
  }
}

type StoredConfig = {
  api_key_env_var: string;
  endpoint_id: string;
  model_name: string;
  timeout?: number;
};

export class RunPodEmbeddingFunction implements IEmbeddingFunction {
  name = "runpod";

  private api_key: string;
  private endpoint_id: string;
  private model: string;
  private timeout: number;
  private runpodApi?: RunPodAPI;

  private api_key_env_var: string;

  constructor({
    runpod_api_key,
    runpod_endpoint_id,
    runpod_model_name,
    runpod_timeout = 300,
    runpod_api_key_env_var = "RUNPOD_API_KEY",
  }: {
    runpod_api_key?: string;
    runpod_endpoint_id: string;
    runpod_model_name: string;
    runpod_timeout?: number;
    runpod_api_key_env_var?: string;
  }) {
    this.api_key_env_var = runpod_api_key_env_var;
    const apiKey = runpod_api_key ?? process.env[runpod_api_key_env_var];
    if (!apiKey) {
      throw new Error(
        `RunPod API key is required. Please provide it in the constructor or set the environment variable ${runpod_api_key_env_var}.`,
      );
    }
    this.api_key = apiKey;

    if (!runpod_endpoint_id || !runpod_endpoint_id.trim()) {
      throw new Error("RunPod endpoint ID is required and cannot be empty.");
    }
    this.endpoint_id = runpod_endpoint_id;

    if (!runpod_model_name || !runpod_model_name.trim()) {
      throw new Error("RunPod model name is required and cannot be empty.");
    }
    this.model = runpod_model_name;

    this.timeout = runpod_timeout;
  }

  private async loadClient() {
    // cache the client
    if (this.runpodApi) return;

    try {
      const { runpodSdk } = await RunPodEmbeddingFunction.import();

      // Validate SDK installation
      if (!runpodSdk || typeof runpodSdk !== "function") {
        throw new Error("Invalid runpod-sdk installation detected");
      }

      this.runpodApi = new RunPodAPIImpl(
        this.api_key,
        this.endpoint_id,
        runpodSdk,
      );
    } catch (error: any) {
      if (error.code === "MODULE_NOT_FOUND") {
        throw new Error(
          "Please install the runpod-sdk package to use the RunPodEmbeddingFunction, e.g. `npm install runpod-sdk`",
        );
      }
      throw error; // Re-throw other errors
    }
  }

  public async generate(texts: string[]): Promise<number[][]> {
    await this.loadClient();

    if (!texts || texts.length === 0) {
      return [];
    }

    return await this.runpodApi!.createEmbedding({
      model: this.model,
      input: texts,
      timeout: this.timeout,
    }).catch((error: any) => {
      throw error;
    });
  }

  /** @ignore */
  static async import(): Promise<{
    runpodSdk: any;
  }> {
    try {
      // @ts-ignore
      const runpodModule = await import("runpod-sdk");
      const runpodSdk = runpodModule.default || runpodModule;
      return { runpodSdk };
    } catch (e: any) {
      if (
        e.code === "ERR_MODULE_NOT_FOUND" ||
        e.message?.includes("Cannot resolve module")
      ) {
        throw new Error(
          "Please install the runpod-sdk package to use the RunPodEmbeddingFunction, e.g. `npm install runpod-sdk`",
        );
      }
      // For other errors, re-throw with more details
      throw new Error(`Failed to import runpod-sdk: ${e.message || String(e)}`);
    }
  }

  buildFromConfig(config: StoredConfig): RunPodEmbeddingFunction {
    return new RunPodEmbeddingFunction({
      runpod_api_key_env_var: config.api_key_env_var,
      runpod_endpoint_id: config.endpoint_id,
      runpod_model_name: config.model_name,
      runpod_timeout: config.timeout,
    });
  }

  getConfig(): StoredConfig {
    return {
      api_key_env_var: this.api_key_env_var,
      endpoint_id: this.endpoint_id,
      model_name: this.model,
      timeout: this.timeout,
    };
  }

  validateConfigUpdate(oldConfig: StoredConfig, newConfig: StoredConfig): void {
    if (oldConfig.model_name !== newConfig.model_name) {
      throw new Error("Cannot change model name.");
    }
    if (oldConfig.endpoint_id !== newConfig.endpoint_id) {
      throw new Error("Cannot change endpoint ID.");
    }
  }

  validateConfig(config: StoredConfig): void {
    validateConfigSchema(config, "runpod");
  }

  defaultSpace(): EmbeddingFunctionSpace {
    return "cosine";
  }

  supportedSpaces(): EmbeddingFunctionSpace[] {
    return ["cosine", "l2", "ip"];
  }
}

File: chromadb/utils/embedding_functions/runpod_embedding_function.py
from chromadb.api.types import Embeddings, Documents, EmbeddingFunction
from typing import List, Dict, Any, Optional, Union
from chromadb.utils.embedding_functions.schemas import validate_config_schema

import os
import numpy as np
import warnings

class RunPodEmbeddingFunction(EmbeddingFunction[Documents]):
    def __init__(
        self,
        endpoint_id: str,
        model_name: str,
        api_key_env_var: str = "RUNPOD_API_KEY",
        api_key: Optional[str] = None,
        timeout: Optional[int] = 300,
    ):
        """
        Initialize the RunPodEmbeddingFunction.
        
        Args:
            endpoint_id (str): The RunPod endpoint ID for your embedding model
            model_name (str): The name of the model to use for embeddings.
            api_key (str, optional): The RunPod API key. If not provided, will use environment variable.
            timeout (int, optional): Timeout in seconds for API requests. Defaults to 300.
        """
        try:
            import runpod
        except ImportError:
            raise ValueError(
                "The runpod python package is not installed. Please install it with `pip install runpod`"
            )

        if api_key is not None:
            warnings.warn(
                "Direct api_key configuration will not be persisted. "
                "Please use environment variables via api_key_env_var for persistent storage.",
                DeprecationWarning,
            )

        self.api_key_env_var = api_key_env_var
        self.api_key = api_key or os.getenv(api_key_env_var)
        if not self.api_key:
            raise ValueError(f"The {api_key_env_var} environment variable is not set.")

        self.endpoint_id = endpoint_id
        self.model_name = model_name
        self.timeout = timeout
        self._runpod = runpod  # Store module reference for later use

    def __call__(self, input: Documents) -> Embeddings:
        """
        Generate embeddings for the given documents.

        Args:
            input: Documents to generate embeddings for.

        Returns:
            Embeddings for the documents.
        """
        # Handle empty input
        if not input:
            return []

        # Ensure input is a list
        if isinstance(input, str):
            input = [input]

        # Set API key before each call to avoid multi-tenant conflicts
        self._runpod.api_key = self.api_key
        endpoint = self._runpod.Endpoint(self.endpoint_id)

        # Submit all requests in parallel first
        run_requests = []
        for document in input:
            input_payload = {
                "input": {
                    "model": self.model_name,
                    "input": document
                }
            }
            run_request = endpoint.run(input_payload)
            run_requests.append(run_request)

        # Collect results
        embeddings = []
        for i, run_request in enumerate(run_requests):
            try:
                status = run_request.status()

                if status in ("FAILED", "CANCELLED", "TIMED_OUT"):
                    raise RuntimeError(f"RunPod endpoint failed with status '{status}': {run_request}")
                elif status == "COMPLETED":
                    output = run_request.output()
                else:
                    output = run_request.output(timeout=self.timeout)

                if output and "data" in output:
                    data_list = output["data"]
                    if len(data_list) > 0 and "embedding" in data_list[0]:
                        embedding = data_list[0]["embedding"]
                    else:
                        raise ValueError(f"No embedding found in response data: {data_list}")
                else:
                    raise ValueError(f"Unexpected output format. Expected 'output.data[0].embedding', got: {output}")

                embeddings.append(np.array(embedding, dtype=np.float32))

            except Exception as e:
                raise RuntimeError(f"Failed to generate embedding for document {i}: {e}")

        return embeddings

    @staticmethod
    def name() -> str:
        """Return the name of this embedding function."""
        return "runpod"

    def default_space(self) -> str:
        """Return the default space for this embedding function."""
        # Most embedding models work best with cosine similarity
        return "cosine"
    
    def supported_spaces(self) -> List[str]:
        """Return the supported spaces for this embedding function."""
        return ["cosine", "l2", "ip"]
    

    def list_available_models(self) -> List[str]:
        """Return the list of models supported by this embedding endpoint."""
        # Set API key before call to avoid multi-tenant conflicts
        self._runpod.api_key = self.api_key
        endpoint = self._runpod.Endpoint(self.endpoint_id)

        models = []
        try:
            run_request = endpoint.run({ "input": { "openai_route": "/v1/models" } })
            status = run_request.status()

            # Get output with timeout handling
            if status != "COMPLETED":
                output = run_request.output(timeout=self.timeout)
            else:
                output = run_request.output()

            if output and "data" in output:
                models = [model["id"] for model in output["data"]]
            else:
                raise ValueError(f"Unexpected output format. Got: {output}")
        except Exception as e:
            raise RuntimeError(f"Failed to list models: {e}")

        return models
        
    @staticmethod
    def build_from_config(config: Dict[str, Any]) -> "EmbeddingFunction[Documents]":
        """Build the embedding function from a configuration dictionary."""
        endpoint_id = config.get("endpoint_id")
        model_name = config.get("model_name")
        api_key_env_var = config.get("api_key_env_var", "RUNPOD_API_KEY")
        timeout = config.get("timeout", 300)

        if endpoint_id is None or model_name is None:
            raise ValueError("endpoint_id and model_name are required in config")

        return RunPodEmbeddingFunction(
            endpoint_id=endpoint_id,
            model_name=model_name,
            api_key_env_var=api_key_env_var,
            timeout=timeout,
        )

    def get_config(self) -> Dict[str, Any]:
        """Get the configuration for this embedding function."""
        return {
            "endpoint_id": self.endpoint_id,
            "model_name": self.model_name,
            "api_key_env_var": self.api_key_env_var,
            "timeout": self.timeout,
            # Note: We don't include api_key for security reasons
        }

    def validate_config_update(
        self, old_config: Dict[str, Any], new_config: Dict[str, Any]
    ) -> None:
        """Validate that a configuration update is allowed."""
        # Prevent changing the model name as it could affect embedding compatibility
        if "model_name" in new_config and new_config["model_name"] != old_config.get("model_name"):
            raise ValueError(
                "The model name cannot be changed after the embedding function has been initialized."
            )
        
        # Prevent changing the endpoint_id as it could affect embedding compatibility
        if "endpoint_id" in new_config and new_config["endpoint_id"] != old_config.get("endpoint_id"):
            raise ValueError(
                "The endpoint_id cannot be changed after the embedding function has been initialized."
            )

    @staticmethod
    def validate_config(config: Dict[str, Any]) -> None:
        """
        Validate the configuration using the JSON schema.

        Args:
            config: Configuration to validate

        Raises:
            ValidationError: If the configuration does not match the schema
        """
        validate_config_schema(config, "runpod")